src/proxy/executor.rs

```diff
@@ -958,6 +914,138 @@ where
+    }
+
+    fn get_blocking_command_timeout(cmd_ctx: &CmdCtx) -> Result<u64, RespVec> {
+        cmd_ctx
+            .get_cmd()
+            .get_command_last_element()
+            .ok_or_else(|| Resp::Error(b"ERR wrong number of arguments".to_vec()))
+            .and_then(|last| {
+                btoi::btou::<u64>(last)
+                    .map_err(|_| Resp::Error(b"ERR invalid timeout argument".to_vec()))
+            })
+    }
+
+    fn transfer_cmd_from_blocking_to_non_blocking(
+        cmd_ctx: &CmdCtx,
+        data_cmd_type: DataCmdType,
+        arg_len: usize,
+        non_blocking_cmd_name: &str,
+    ) -> Result<NonBlockingCommandsWithKey, RespVec> {
+        let mut cmds = vec![];
+        use DataCmdType::*;
+        match data_cmd_type {
+            BLPOP | BRPOP | BZPOPMIN | BZPOPMAX => {
+                // exclude the timeout argument
+                for i in 1..(arg_len - 1) {
+                    let key = match cmd_ctx.get_cmd().get_command_element(i) {
+                        None => break, // invalid state
+                        Some(key) => key.to_vec(),
+                    };
+                    let non_blocking_cmd =
+                        vec![non_blocking_cmd_name.to_string().into_bytes(), key.clone()];
+                    let arr: Vec<RespVec> = non_blocking_cmd
+                        .into_iter()
+                        .map(|s| Resp::Bulk(BulkStr::Str(s)))
+                        .collect();
+                    let resp = Resp::Arr(Array::Arr(arr));
+                    cmds.push((key, resp));
+                }
+            }
+            BRPOPLPUSH => {
+                let mut resp = cmd_ctx.get_cmd().get_resp_slice().map(|b| b.to_vec());
+                change_bulk_array_element(
+                    &mut resp,
+                    0,
+                    non_blocking_cmd_name.to_string().into_bytes(),
+                );
+                if let Resp::Arr(Array::Arr(ref mut resps)) = resp {
+                    resps.pop(); // pop out the timeout argument
+                }
+                // BRPOPLPUSH does not need to care about key.
+                cmds.push((vec![], resp));
+            }
+            _ => return Err(Resp::Error(b"ERR unsupported blocking command".to_vec())),
+        }
+        Ok(cmds)
+    }
+
+    fn adjust_lrpop_response(resp: RespVec, key: Vec<u8>) -> RespVec {
+        match resp {
+            Resp::Bulk(BulkStr::Nil) => {
+                // BLPOP, BRPOP need to change resposne to Array::Nil.
+                return Resp::Arr(Array::Nil);
+            }
+            Resp::Bulk(BulkStr::Str(s)) => Resp::Arr(Array::Arr(vec![
+                Resp::Bulk(BulkStr::Str(key)),
+                Resp::Bulk(BulkStr::Str(s)),
+            ])),
+            _ => resp,
+        }
+    }
+
+    fn adjust_zpop_response(resp: RespVec, key: Vec<u8>) -> RespVec {
+        match resp {
+            Resp::Arr(Array::Arr(arr)) if arr.len() == 0 => Resp::Bulk(BulkStr::Nil),
```

Comment:
The [docs](https://redis.io/commands/bzpopmin) says it's a nil multi bulk:
```
Array reply: specifically:

A nil multi-bulk when no element could be popped and the timeout expired.
```

which is `Resp::Arr(Array::Nil)`